### PR TITLE
Catalog of Things App

### PR DESCRIPTION
**Pull request by Tayyab-Kamboh**
**The following implementations are being introduced:**

- Used the UML class diagram provided to create the required classes and associations between classes in separate files.

- Followed the diagram to create the methods within the respective classes.

- Added unit tests for all implemented methods

- Created a main.rb file to serve as the entry point of our console application

- At the start, the console app does the following:

- Presents the user with a list of options to perform.

- Lets the user choose an option.

- If needed, ask for parameters for the option.

- The app presents an option to quit.

-  The options presented are as follows:

- List all books

- List all music albums

- List all movies

- List of games

- List all genres

- List all labels

- List all authors

- Add a book

- Add a music album

- Add a game

-  All data from the book, music album, and game is preserved in files using  .json format

- Link to project video: https://drive.google.com/file/d/1DkOikLbC6yEkCnHbQ_eZkuCBj7mzogs8/view?usp=sharing
  NOTE: Our partner Abbas could not make it to the video call due to some internet issues , so me and Otega described his 
  part of the project as well.

Thank You for Reviewing ✨

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    app.add_music_album
+  when 9
+    app.add_game
+  when 10
+    puts 'Thank you for using the app!'
+    break
+  else
+    puts 'Invalid Option'
+  end
+end
```
Comment (muneebulrehman) [inline main.rb]: After running this file, I am getting some unexpected behavior. Here you can see in the screenshot. Kindly fix this issue.
image

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    @publisher = publisher
+    @cover_state = cover_state
+  end
+
+  private
+
+  def can_be_archived?
+    super || @cover_state == 'bad'
+  end
+end
```
Comment (muneebulrehman) [inline book.rb]: The book should have information on all the association classes that you described in the Item class. ie., `Label`, `Author`, and `Genre`. But in your case, the Book only asks for `Label` and not the `Author` and `Genre`. As you can see in the class diagram, Item has a 1-many association with the rest of the classes and hence that should be included in the Book as it inherits from the Item. Kindly fix this issue.

File: game.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    super(id, publish_date)
+    @multiplayer = multiplayer
+    @last_played_at = last_played_at
+  end
+
+  def can_be_archived?
+    is_expired = Time.now.year - Date.parse(@last_played_at).year > 2
+    super && is_expired
+  end
+end
```
Comment (muneebulrehman) [inline game.rb]: The game should have information on all the association classes that you described in the Item class. ie., `Label`, `Author`, and `Genre`. But in your case, the game only asks for `Author` and not the `Label` and `Genre`. As you can see in the class diagram, Item has a 1-many association with the rest of the classes and hence that should be included in the Game as it inherits from the Item. Kindly fix this issue.

File: music_album.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    super(id, publish_date)
+    @on_spotify = on_spotify
+  end
+
+  private
+
+  def can_be_archived?
+    super && @on_spotify
+  end
+end
```
Comment (muneebulrehman) [inline music_album.rb]: The MusicAlbum should have information on all the association classes that you described in the Item class. ie., `Label`, `Author`, and `Genre`. But in your case, the MusicAlbum only asks for `Genre` and not the `Author` and `Label`. As you can see in the class diagram, Item has a 1-many association with the rest of the classes and hence that should be included in the MusicAlbum as it inherits from the Item. Kindly fix this issue.

File: app.rb [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  end
+
+  def list_games
+    return puts 'No games found' if @preserve.games.empty?
+
+    @preserve.games.each_with_index do |game, index|
+      puts "#{index + 1} Publish Date: #{game.publish_date},
+      Multiplayer: #{game.multiplayer}, Last played at: #{game.last_played_at}"
+    end
+  end
```
Comment (muneebulrehman) [inline app.rb]: Whether listing books, music albums, or games, it should display all the information about that particular item. That includes information about the label, author, and genre. However, in your case, it displays very limited information which is not very useful and doesn't give any real information about the item. Kindly fix this issue. Eg., have a look at the screenshot below, when i list music albums.

image

File: spec/book_spec.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    it 'should return publisher name' do
+      book = Book.new(nil, Date.parse('1999-04-07'), 'Birhan and Selam', 'bad')
+      expect(book.publisher).to eq 'Birhan and Selam'
+    end
+
+    it 'should return cover state' do
+      book = Book.new(nil, Date.parse('1999-04-07'), 'Birhan and Selam', 'bad')
+      expect(book.cover_state).to eq 'bad'
+    end
+  end
```
Comment (muneebulrehman) [inline spec/book_spec.rb:15]: Kindly test for the `can_be_archived?` method in the book class. The test for this particular method is missing.

File: item.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
+class Item
```
Comment (muneebulrehman) [inline item.rb:1]: Kindly add the tests for the `Item` class. Tests should cover all the methods in this class.

**Review by muneebulrehman — CHANGES_REQUESTED** [review]
Hi team, 

Good job so far! I see you put a lot of effort into this project. However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- No linter errors
- Descriptive PR
- Clean code
- Great job with the project so far 👍


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: main.rb (hunk 1 above)
Comment (Tayyab-Kamboh) [reply main.rb]: Kindly elaborate what behaviour exactly? Thank you

Comment (muneebulrehman) [reply main.rb]: Kindly check what is printed in the console after I run `main.rb` file. It probably is displaying music album data which should not be happening.

**Review by DeliceLydia — APPROVED** [review]
## Hi Team, 

In my opinion, your project is ready for presentation!  There is nothing else to say other than... it's time to merge it :shipit: 


### Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Pull request merged**